src/Build/Graph/ProjectGraph.cs

```diff
@@ -60,14 +65,103 @@ public ProjectGraph(string entryProjectFile)
         /// </summary>
         public IReadOnlyCollection<ProjectGraphNode> ProjectNodes { get; }
 
+        /// <summary>
+        /// Gets the target list to be executed for every project in the graph, given a particular target list for the entry project.
+        /// This method uses the ProjectReferenceTargets items to determine the targets to run per node. The results can then be used
+        /// to start building each project individually, assuming a given project is built after its references.
+        /// </summary>
+        /// <param name="entryProjectTargets">The target list for the entry project. May be null or empty, in which case the project's default targets will be used.</param>
+        /// <returns>A dictionary containing the target list for each node.</returns>
+        public IReadOnlyDictionary<ProjectGraphNode, string[]> GetTargetLists(string[] entryProjectTargets)
+        {
+            List<string> entryTargets = entryProjectTargets == null || entryProjectTargets.Length == 0
+                ? HydrateDefaultTargets(EntryProjectNode.Project, new List<string> { DefaultTargetsMarker })
```

Comment:
Rather than hydrate, you can just add the default targets directly from the project instance.

Reply:
And clone it.

---
In reply to: [219355209](https://github.com/Microsoft/msbuild/pull/3769#discussion_r219355209) [](ancestors = 219355209)

Reply:
Good call.